PanelChrome: Refactor and refine items next to title

axelavargas opened the pull request:
**What is this feature?**

Part of https://github.com/grafana/grafana/issues/56163 

**Why do we need this feature?**

Connect new panel header title items  to PanelChrome (Design [here](https://www.figma.com/file/jdTCXlq75p55itRRSaerFx/Panel-header-and-menu?node-id=2672%3A235&t=DjT1N7RKflKEv2x2-1))  - behind `newPanelChromeUI` feature flag 

Extend the properties that `PanelChrome` support, we added `description` as standard prop and started using `titleItems`  for edge cases (links, notices, timeshift and alerts)

- [x] Create new `PanelHeaderTitleItems` in core grafana that contains links, notices, timeshift and alerts
- [x] Refactor `PanelChrome`, now receives standard props 
	- [x]  Implement PanelNotices
	- [x]  Implement  Panel Links
	- [x]  Implement Description,
	- [x] A11y: Make sure the description, panel links, and health is accessible via the keyboard 
	- [x] CSS: Each component should have a height and width of 32px, centered, for timeshift width is not mandatory
	- [x] Remove `section` and get rid of global variables in `PanelStateWrapper`
	- [x] Refactor `titleItems` to be a list of react nodes instead of a render prop
	- [x] CSS: Fix overflow for panel header
	- [x] CSS: Implement alerting colors in panelTitleItems instead of relying on global CSS
	- [x] A11y: Refactor links, panel notices, alerts, and timeshift to don't use `ToolbarButton` as they are not interactive. 
	- Add unit test for new components --- follow up PR

**Which issue(s) does this PR fix?**:

Fixes #59080

**Special notes for your reviewer**:

- Levitate might mark the changes on this PR as a breaking change, this is not relevant because `titleItems` was added in another PR (targeting 9.4 as well, not released yet, not affecting plugin developers)

axelavargas committed:
Playground, extract relevant logic to titleItems component, render in
panel chrome

axelavargas committed:
Use title items as render call
Make Menu for panel links to be keyboard accessible
Use IconButton for description and health

axelavargas committed:
Add panelHeaderNotices and timeshift to title items

torkelo commented on packages/grafana-ui/src/components/PanelChrome/PanelChrome.tsx:
```diff
@@ -30,7 +16,7 @@ export interface PanelChromeProps {
   padding?: PanelPadding;
   title?: string;
-  titleItems?: PanelChromeInfoState[];
+  titleItems?: (innerWidth: number, innerHeight: number) => ReactNode;
```
why does this need innerWidth and innerHieght? (and innerWidth and innerHeight of what, header, panel content?,)

torkelo replied:
but I think it could be a good idea to make this a more extensible hook, I think PanelChrome might need to be very generic and support a lot of different flavors / features (as we explore more use cases from Scenes framework)

polibb replied:
>why does this need innerWidth and innerHieght?

The idea is that the space _each_ "item" is allowed is determined by the `PanelChrome` and passed down to the component responsible for setting up the items themselves (in our case the `PanelHeaderTitleItems`). The innerWidth and innerHeight here are probably better named infoNodeHeight and infoNodeWidth or something of the kind. I can see they are not yet used in PanelHeaderTitleItems but will set the 32x32 px square we have around each icon:
![image](https://user-images.githubusercontent.com/13227501/209099949-b49fafeb-0a68-4f40-8b71-ce47497ecb47.png)

torkelo replied:
Not sure I understand still , seems unnecessary?

torkelo replied:
I can't see it being used in https://github.com/grafana/grafana/blob/cf99a137932cd1126b166a3bb5e449c92dfece6f/public/app/features/dashboard/dashgrid/PanelHeader/PanelHeaderTitleItems.tsx

axelavargas replied:
Indeed is not being used in the current implementation and is not needed.

When we did the analysis, we thought we could bring consistency to the rendering of the `titleItems` by offering the width and height, like saying "32px of height is what your items should have". However, when I was implementing this, I noticed how complicated it is to do that. For example, I had to pass the `innerWidth` and `innerHeight` down to 3 different components in order to render the Panel notices with those dimensions (PanelHeaderTitleItems --> PanelHeaderNotices --> PanelHeaderNotice).

I would say that this was a failed attempt to bring control and consistency to the approach of having `title items` as a render call (the most significant disadvantage of this approach is that we don't have control over what the developer will place in it, but it is also a big advantage considering we are creating a component that can be used outside of the Grafana core domain).

I will remove the exposure of these dimensions and add the task of adding to the PanelChrome MDX documentation that `title items at least should have 32px as their height`.

dprokop replied:
> I will remove the exposure of these dimensions and add the task of adding to the PanelChrome MDX documentation that title items at least should have 32px as their height.

IMO the sizing of the title items should be set via wrapper html element that renders it - why would someone using the API have to set the dimensions of an action item? It's an easy path for creating inconsistencies in the panel chrome.

axelavargas replied:
Thanks a lot for the feedback 🙌🏾 , after our discussion I have made changes that tackle this point:

- `PanelChrome` now supports standard props such as `description, links, and panelNotices` 
- `titleItems` is now an array of React nodes that will render edge cases like `alerts and timeshift`

torkelo commented on public/app/features/dashboard/dashgrid/PanelHeader/PanelHeaderTitleItems.tsx:
```diff
@@ -0,0 +1,83 @@
+  panelDescription?: string;
+  alertState?: string;
+  scopedVars?: ScopedVars;
+  replaceVariables: (value: string, extraVars: ScopedVars | undefined, format?: string | Function) => string;
```
```suggestion
  replaceVariables: InterpolateFunction;
```

polibb committed:
use InterpolateFunction as replaceVariables type

Co-authored-by: Torkel Ödegaard <torkel@grafana.com>

Commit:
snall fixes

axelavargas committed:
Styling issues fixes:

- Make PanelHeaderNotices keyboard accessible
- Make timeshift, description, panel links with the right width, height
- Add new section to wrapper PanelChrome in order to get alerting
  styles

torkelo commented on public/app/features/dashboard/dashgrid/PanelHeader/PanelHeaderTitleItems.tsx:
```diff
@@ -19,8 +28,9 @@ export interface Props {
+  // description
   const rawDescription = panelDescription || '';
   const descriptionMarkdown = getTemplateSrv().replace(rawDescription, scopedVars);
   const description = renderMarkdown(descriptionMarkdown);
```
don't render markdown on every header render, this needs to be lazy (when you hover over it / show tooltip )

axelavargas replied:
good catch, I will make the changes :)

torkelo commented on public/app/features/dashboard/dashgrid/PanelStateWrapper.tsx:
```diff
@@ -582,33 +584,68 @@ export class PanelStateWrapper extends PureComponent<Props, State> {
     });
 
     // for new panel header design
+    const containerClassNamesNewPanelChrome = classNames({
+      'panel-container--absolute': isSoloRoute(locationService.getLocation().pathname),
+      'panel-container--transparent': transparent,
+      'panel-container--no-title': this.hasOverlayHeader(),
+      [`panel-alert-state--${alertState}`]: alertState !== undefined,
+    });
+
```
??  the point of the new PaneChrome is to replace all these

axelavargas replied:
I added them because I was not getting the right colors for the alert states(firing: red, pending: yellow, ok: green) and I did not want to implement that logic on this PR. I will check the code handling the alerts and see how much work would require.

torkelo replied:
Alert state is a title item right?  Not sure why a class in an outer element affects the color of a title item

axelavargas replied:
yes it's a title item, as far as I see we are giving the colors using some global alert CSS https://github.com/grafana/grafana/blob/23956557d8c6a119b7de5be5c42024e29634d002/public/sass/pages/_alerting.scss#L47

torkelo replied:
but add the correct class in https://github.com/grafana/grafana/pull/60514/files#diff-72fa0ae4d8ddfb483e3ccb61c7bbc3e17ed388182132ebe6a89446414a111daeR19 

or color it via emotion styles in PanelHeaderTitleItems.tsx 

Not by adding these legacy classes to wrapper div here, sort of defeats the purpose of PanelChrome :)

polibb replied:
@torkelo, I'll add the correct classes for alerts coloring in `PanelHeaderTitleItems` after I add them to `GrafanaTheme2` in order to follow the dark/light mode differences. 
This will take care of `[panel-alert-state--${alertState}]: alertState !== undefined`.

`'panel-container--no-title': this.hasOverlayHeader()` will be removed entirely, it is handled by @axelavargas's implementation of `timeshift` prop.

This way of styling the panel because of its different properties has to go indeed. However we are aiming to finish up the functionality of title items, so that we can implement the **hovering header** next.
So `'panel-container--absolute': isSoloRoute(locationService.getLocation().pathname)` and `'panel-container--transparent': transparent` will stay in this legacy form for now and will be fixed in a separate PR.

torkelo replied:
@polibb why do we need 'panel-container--absolute': isSoloRoute(locationService.getLocation().pathname) and 'panel-container--transparent': ? can't we skip the section div for now and add transparent display mode and changes for solo route in follow-ups instead?

torkelo replied:
> after I add them to GrafanaTheme2 in order to follow the dark/light mode differences.

What do you need to add to GrafanaTheme2? I don't see any need to add any new things to theme (it already has theme.colors.error theme.colors.success)

polibb replied:
>What do you need to add to GrafanaTheme2? I don't see any need to add any new things to theme (it already has theme.colors.error theme.colors.success)

Yes, sorry, I spoke before checking if we have all of them covered. We do 😊

polibb replied:
>can't we skip the section div for now and add transparent display mode to PanelChrome and changes for solo route in follow-ups instead?

Yes, pushed changes to remove any global styles and will leave transparent and solo panel functionalities for followups.

torkelo commented on public/app/features/dashboard/dashgrid/PanelStateWrapper.tsx:
```diff
@@ -582,33 +584,68 @@ export class PanelStateWrapper extends PureComponent<Props, State> {
-            </>
-          )}
-        </PanelChrome>
+        <section
+          className={containerClassNamesNewPanelChrome}
+          aria-label={selectors.components.Panels.Panel.containerByTitle(panel.title)}
+        >
```
this section should not be needed, aria-label can be moved to PanelChrome

polibb replied:
Yes, that makes sense, aria-label is now in PanelChrome. We'll remove the `section` [when we implement](https://github.com/grafana/grafana/pull/60514#discussion_r1067980557) better ways to handle _transparency_ and _embedded solo panel_ in a next PR.

torkelo replied:
@polibb would just remove it in this PR as it adds no value in its current state, but leave it to you to decide

torkelo replied:
> We'll remove the section https://github.com/grafana/grafana/pull/60514#discussion_r1067980557 better ways to handle transparency and embedded solo panel in a next PR.

'panel-container--absolute' and  'panel-container--transparent' have no impact on PanelChrome so just not sure why it was added this wrapper, these classes only impact the old table styles which the new PanelChrome don't use :)

polibb replied:
The new panel will not be able to be shared in an embedded form (absolute positioning) and will not have the transparent (background) option?

torkelo replied:
@polibb of course it will, but just saying that those classes have no effect on this new PanelChrome so not sure why they wrap the new PanelChrome

dprokop commented:
> We decided to pass down titleItems as a render prop instead of the constraint array PanelChromeInfoState because we wanted to support rendering panel links and panel header notices.

@axelavargas can you elaborate more on this? Want to understand the need for render prop here. I think I get it for panel links, as they are generated in a "lazy" way (which BTW can be achieved without render prop, but via -  i.e. Tooltip render prop - similar thing is done for the Datalinks context menu). But don't get the header notices.

dprokop commented on public/app/features/dashboard/dashgrid/PanelHeader/PanelHeaderTitleItems.tsx:
```diff
@@ -0,0 +1,124 @@
+  const description = renderMarkdown(descriptionMarkdown);
+
+  // panel links
+  const panelLinks = links && links.getLinks(replaceVariables);
```
Similar to comment from @torkelo about markdown - the links can be rendered when necessary, i.e. dropdown is opened. A similar thing is implemented in the Data links context menu.

axelavargas committed:
Implement description in panelChrome

axelavargas committed:
move panel links to PanelChrome

axelavargas committed:
Move panel header notices to PanelChrome

axelavargas committed:
Add missing PanelNotices component and restore PanelHeaderNotice changes
keep changes from main branch

axelavargas committed:
Refactor titleItems now is a list of react nodes

axelavargas committed:
Merge branch 'main' into axelav/panel-header-title-items

dprokop commented on packages/grafana-ui/src/components/PanelChrome/PanelChrome.tsx:
```diff
@@ -124,6 +125,18 @@ export function PanelChrome({
           </h6>
         )}
 
+        {panelNotices && typeof panelNotices.getPanelNotices === 'function' && (
```
Think the `typeof panelNotices.getPanelNotices === 'function'` is not necessary here since it cannot be anything else?

dprokop commented on packages/grafana-ui/src/components/PanelChrome/PanelChrome.tsx:
```diff
@@ -124,6 +125,18 @@ export function PanelChrome({
+
+        <PanelLinks links={links} />
+
+        <div className={styles.titleItems} data-testid="title-items">
```
This div should be rendered only when `titleItems` are provided.

dprokop commented on packages/grafana-ui/src/components/PanelChrome/PanelChrome.tsx:
```diff
@@ -124,6 +125,18 @@ export function PanelChrome({
+        <PanelLinks links={links} />
+
+        <div className={styles.titleItems} data-testid="title-items">
+          {itemsRenderer(titleItems, (item) => item)}
```
May be my personal preference, but I don't see this renderer bringing more value than a regular `map` :)

polibb replied:
This is indeed how the PanelChrome used to render the set of leftItems before we started making changes for the new design. 
Maybe you're right that the value is not exceedingly greater, but `React.Children.toArray()` returns `Array<Exclude<ReactNode, boolean | null | undefined>>` which at least makes sure that we are only rendering valid ReactNodes here.
On the other hand, `titleItems` is being put through `toArray()` and then `filter()` method, whereas otherwise it would just go through one `map()`.

axelavargas committed:
Apply PR suggestions:
- Fix header height when title is not present
- Add aria label to Panel Chrome
- Add contidion to prevent rendering div if title items is not present

torkelo commented on packages/grafana-ui/src/components/PanelChrome/PanelChrome.tsx:
```diff
@@ -39,8 +35,11 @@ export interface PanelChromeProps {
-  menu?: ReactElement;
+  description?: string | (() => string);
+  links?: () => Array<LinkModel<PanelModel>>;
+  panelNotices?: Notices;
```
I don't think this interface makes much sense, think a simple react node | function return react node is better here (title items should suffice )

torkelo reviewed (commented):
Not sure we need the notices, and links when we have titleItems on PanelChrome pros already

torkelo commented on packages/grafana-ui/src/components/PanelChrome/PanelChrome.tsx:
```diff
@@ -124,6 +139,18 @@ export function PanelChrome({
           </h6>
         )}
 
+        {panelNotices && <PanelNotices notices={panelNotices.getPanelNotices()} onClick={panelNotices.onClick} />}
+
+        <PanelDescription description={description} />
+
+        <PanelLinks links={links} />
+
+        {titleItems && (
+          <div className={styles.titleItems} data-testid="title-items">
```
I think panel notices and links could just use regular title items, don't think we need to treat them differently and would make this component a bit simpler

torkelo commented on packages/grafana-ui/src/components/PanelChrome/PanelLinks.tsx:
```diff
@@ -0,0 +1,41 @@
+  links: (() => LinkModel[]) | undefined;
+}
+
+export function PanelLinks({ links }: Props) {
```
Panel links as drop-down menu is a change from the tooltips we use today, is there a reason to change them from tooltips (tooltips don't require clicks)

axelavargas replied:
Yes, I want to make the links accessible via the keyboard, so I decided to use the `Dropdown` component. Unfortunately, this component only supports `click` to display the menu, there is no `hover` option :(.

torkelo replied:
I think each link should be it's own icon and be a href

axelavargas replied:
🤔 do you mean that if we configure 3 panel links, the panel header should have 3 of these icons?

![image](https://user-images.githubusercontent.com/239999/211762508-94a08ef0-f119-43b2-bafb-13449669e06a.png)

If yes, I believe this can make very crowded the panel header, as we don't have a limit on panel links, also it is not aligned with the current design.

torkelo replied:
@axelavargas yes, that is one of the benefits of this new design is to give more space and make links more accessible so you can access them without hovering over a tooltip. 

But true, you can have multiple. And if more than one maybe collapsing them into a menu is better

axelavargas replied:
I have followed the approach you recommended @dprokop , it looks way better 🙌🏾, [here relevant commit](https://github.com/grafana/grafana/pull/60514/commits/e445a727e25dc574ea2bb955050234d4894a890c)

torkelo commented on packages/grafana-ui/src/components/PanelChrome/PanelNotices.tsx:
```diff
@@ -0,0 +1,52 @@
+
+    if (notice.inspect && onClick) {
+      return (
+        <ToolbarButton
```
Toolbar button changes look based on if topnav is enabled , not sure that good choice for panel header.  

Maybe a regular button with fill text could work, hmmm

axelavargas replied:
🤔 I did not know we do that. I will double-check this point.

axelavargas replied:
@torkelo I didn't see changes in the `ToolbarButton` when I enabled the `topnav`, am I missing something?

https://user-images.githubusercontent.com/239999/211595397-5df5fa34-2b07-4c5d-a09b-6c807826ce20.mp4

torkelo replied:
@axelavargas strange, the button should have a border when topnav is disabled

axelavargas replied:
Oh right! I added a CSS class to remove that border, in that way I was aligned with the design.

torkelo replied:
hm.. maybe stupid, let's keep ToolbarButton for now (except for timeshift, as it is just a info text and not button right now)

torkelo commented on public/app/features/dashboard/dashgrid/PanelHeader/PanelHeaderTitleItems.tsx:
```diff
@@ -0,0 +1,58 @@
+
+  const timeshift = (
+    <>
+      <ToolbarButton
```
Timeshift info is not a button really (no action) , but I think long term it will be a button that opens a modal so maybe it's fine so long remove cursor and focus in the meantime

torkelo replied:
I think it also should not be blue (that should be reserved for links), thoughts @Ijin08 ?

axelavargas commented:
Hey @torkelo 😄 thanks for the feedback, I will give some missing context to address some of your comments related to having extra props such as links and panel notices.

### Context: Architecture Session with @dprokop, @ivanortegaalba and @polibb 
 
Last week we had an architecture session and we found drawbacks to the old approach of passing  `titleItems` as a react node/function and having an API that was too open.

-  Developers can put **anything** in the `titleItems`  ---> This can lead to inconsistencies across Grafana panels, and in a way disregard the work we did in the new designs.
-  API is not declarative and can be hard to use for standard props ---> For example, adding a `description` to the panel should not require passing a React node where the developer needs to take care of having the right components, styles, etc...

### New Approach
We decided to go towards a declarative API for props that can be considered standard in the context of a panel (`description, links, panel notices`), and for other props that can be considered edge cases such as timeshift and alerts we offered the `titleItems`. 

#### Why panelNotices is in the standard props?

We needed to keep a specific order[ provided in the design](https://www.figma.com/file/jdTCXlq75p55itRRSaerFx/Panel-header-and-menu?node-id=2780%3A4150&t=Uu4m6RAUZJJNLUy7-0)  and panel notices need to be next to the title  [discussed here](https://raintank-corp.slack.com/archives/C03MCTKAJA0/p1670338162083809?thread_ts=1670331172.095119&cid=C03MCTKAJA0)

![image](https://user-images.githubusercontent.com/239999/211540654-4024ca10-d98f-419b-be9c-3877d6deae9f.png)

Let me know if this helps answer your comments :), happy to jump on a call 😄

Commit:
Merge branch 'axelav/panel-header-title-items' of https://github.com/grafana/grafana into axelav/panel-header-title-items

axelavargas committed:
Fix streaming icon not showing on storybook

torkelo commented:
@axelavargas the use case of this panel is mostly going to be internal , and we will need a lot of flexibility in this panel for the use case from scenes. And since you already have a generic title items the others seem superfluous (maybe expect for description).

A agree with a declarative api but then I would try to find a more generic higher level concept that can unite all the title items. The current interfaces feel a bit messy with objects that have functions that return links etc. 


We will very likely evolve how and where we show links and time range. Think it could be better if the details for how that is handled be outside PanelChrome.

But I am torn to be honest, will try to spend some time and think about this.

torkelo commented on packages/grafana-ui/src/components/PanelChrome/PanelChrome.tsx:
```diff
@@ -1,35 +1,31 @@
-  variant?: IconButtonVariant;
-  onClick?: () => void;
+interface Notices {
+  getPanelNotices: () => QueryResultMetaNotice[];
```
Why the object and function here? Why not use the PanelStateInfo or TitleItem type , mainly wonder why we are passing an object with function that returns items instead of just passing the items.

Ideally we should pass a simple array of PanelChromeNotice type

axelavargas replied:
🤔 that's a good point, I remember following the pattern of passing a function down that was going to be called whenever we did hover on icons (i.e description, or panel links) to get variables interpolated, but double checking, in the case of panel notices this is not necessary. 

I will pass down the list of notices instead 😄

torkelo commented on packages/grafana-ui/src/components/PanelChrome/PanelChrome.tsx:
```diff
@@ -81,7 +83,18 @@ export function PanelChrome({
+  //
+  // Backwards compatibility for having a designated space for the header
+
+  const hasHeader =
```
I thought we decided this logic was going to be outside PanelChrome and we instead pass a prop hoverHeader, that way we can have a panel with title and description/links with no header (shown on hover above it )

But if that change is for a future PR where the hover header is added , if so ignore comment :)

But I see we have the prop named hoverHeader so a bit confused

axelavargas replied:
Yes, `hoverHeader` feature will be tackled on a different PR. We added this logic here to:

1. Keep backward compatibility with the current `PanelChrome`:
	- When `title` and `leftItems` are both undefined, we do NOT have a header.
	- `When `title` is undefined, but we have `leftItems`, then we have a header. 
	
	In this scenario, we can't rely on `hoverHeader` because by default this prop is `false`

2. Put always a header until we implement the `hoverHeader`, because we are targeting this PR to 9.4 and we are not sure `hoverHeader` will be ready for that time,.

torkelo commented on packages/grafana-ui/src/components/PanelChrome/PanelChrome.tsx:
```diff
@@ -288,5 +295,11 @@ const getStyles = (theme: GrafanaTheme2) => {
+      display: 'flex',
+      alignItems: 'center',
+      overflow: 'hidden',
+      padding: theme.spacing(1),
```
this padding I think should be removed, each item has what ever padding it needs?

It looks like this padding extends outside the bounds of the header:
<img width="402" alt="Screenshot 2023-01-10 at 17 41 05" src="https://user-images.githubusercontent.com/10999/211623588-e280937a-b2a6-43f1-b566-ee3a18515e97.png">

axelavargas replied:
Yes, I added this extra padding to prevent the focus ring from disappearing when the panel header gets smaller and the `overflow: hidden` from the title items kicks in, I didn't find a better solution to solve this problem, I saw this issue from [w3c](https://github.com/w3c/csswg-drafts/issues/3417).

https://user-images.githubusercontent.com/239999/211778870-27286ed6-0ac4-41fb-86f2-2a6cd4a565e5.mp4

torkelo replied:
ah, interesting. well if the padding does not cause issues then I guess it's fine, just looked strange when it extended the bounds :) 

There lack of spacing / padding (or gap) between title and the first title item (or description)

dprokop commented:
> @axelavargas the use case of this panel is mostly going to be internal , and we will need a lot of flexibility in this panel for the use case from scenes. And since you already have a generic title items the others seem superfluous (maybe expect for description).
> 
> A agree with a declarative api but then I would try to find a more generic higher level concept that can unite all the title items. The current interfaces feel a bit messy with objects that have functions that return links etc.
> 
> We will very likely evolve how and where we show links and time range. Think it could be better if the details for how that is handled be outside PanelChrome.
> 
> But I am torn to be honest, will try to spend some time and think about this.

@torkelo thanks for the feedback - I understand your concerns. There are a few things we discussed tho, consistency across Grafana being the most relevant one. Regardless of the application of the PanelChrome (name it app, dashboard, or scene), certain things should always appear in the same place; in particular the description, links and notices (though this one is such a non-standard one that I'm on a fence too). With standardized API we are guaranteeing the UI consistency. Now, since customization will probably come into the picture, I think the still present `titleItems` property will allow us to add custom controls, i.e. in scenes. 

Regarding the evolution of how and where the links or time range is shown. With links/description - a standard option for a panel - this will be relatively easy to change given a standard prop, as then we are delegating the rendering of the layout of those to the layout-responsible component which is PanelChrome. I would not expect nor wish for those elements' look and feel to be provided via an external component that can come with any styling. 

Given the above, I think standardizing some basic props and providing an extension point for others (`titleItems`) is giving us a fair amount of flexibility for other applications of the new PanelChrome.

axelavargas commented on public/app/features/dashboard/dashgrid/PanelStateWrapper.tsx:
```diff
@@ -586,39 +612,62 @@ export class PanelStateWrapper extends PureComponent<Props, State> {
+            }}
+            description={!!panel.description ? this.onShowPanelDescription : undefined}
+            links={panel.links && panel.links?.length > 0 ? this.onShowPanelLinks : undefined}
+            panelNotices={{
+              getPanelNotices: () => getPanelQueryNotices({ frames: data.series }),
+              onClick: (e: React.SyntheticEvent, tab: string) => this.onOpenInspector(e, tab),
+            }}
+            titleItems={titleItems}
```
I checked with Patrick and because of the nature of PanelNotices (very edge case), it is not important to have this next to the title. I will move these inside `titleItems` and we can have a cleaner API.

torkelo commented:
@dprokop I agree with the consistency goal but having multiple props that can do the same thing is also a bit confusing.

For links I can see we have a few different ways we show them.
* Turn the title into a link
* Have an icon link (that is a real link)
* Have an icon that opens a dropdown menu for multiple links
* Show a full-text link
* Show a full-text link in footer (PanelChrome only has header right now but could see a world where we add a footer and to enable showing description and links there)

So not sure a single links prop on PanelChrome is going to be enough. And if it's a function that returns an array we will always have to call it to know how many links there are to know if we should show a single real link or a menu (Which defeats the purpose of passing a function). 

So the question is how to enable a flexible future for PanelChrome with sensible props, I find it really hard :)  I would also try to limit passing props via sub objects (status, notices),  passing props via sub-objects make breaks props memoization. I think it's recommended to use flat props as much as possible. 

The thing with the current props is it feels a bit fragmented and complex (esp notices), PanelTitleItems get's passed data and we extract alert state and show that (and timeshift), but notices extracted from data is shown very differently by passing a function directly to PanelChrome that returns query metadata items..

I think ttile, description, titleItems, statusMessage, statusOnClick, should be enough for now. Maybe we can promote more to core props if we see need to share implementation between uses of PanelChrome and inconsistencies that we want to address?

axelavargas committed:
Remove panel notices from standard panel props and fix a11y issue in
panelNotice component

axelavargas committed:
remove unused code related to panel notices

dprokop commented:
> And if it's a function that returns an array we will always have to call it to know how many links there are to know if we should show a single real link or a menu (Which defeats the purpose of passing a function).

This is a good argument for having those passed via `titleItems`, indeed, as we can use the `panel.links` outside of the PanelChrome to decide what to render, and provide the lazy interpolation for drop-down scenario. I agree. cc @axelavargas 

When it comes to Notices - we decided to move them to `titleItems` too. Our thinking here was around the designed order of things, but we talked with @Ijin08 this morning and came to the conclusion, that they don't have to appear as the first item after the panel title - which was our main motivation to standardize notices.

axelavargas committed:
Remove links from  PanelChrome, now lives in PanelHeaderTitleItems

axelavargas committed:
Move PanelLinks component to core

axelavargas committed:
Fix linting issues

axelavargas committed:
Refactor PanelDescription, don't use ToolbarButtos as description is not
a button, make the tooltip focusable

axelavargas committed:
Remove unused imports

Commit:
Merge branch 'main' into axelav/panel-header-title-items

Commit:
Merge branch 'axelav/panel-header-title-items' of https://github.com/grafana/grafana into axelav/panel-header-title-items

axelavargas committed:
Refactor Timeshift, Alerts and PanelNotice to use the right html
sematics, they should be keyboard accessible but they are not a button.

Commit:
remove legacy styling; color alerting inside PanelHeaderTitleItems; remove wrapper around PanelChrome

axelavargas committed:
Refactor PanelLinks component, follow the same approach as links in
Table component

torkelo commented on public/app/features/dashboard/dashgrid/PanelHeader/PanelHeaderTitleItems.tsx:
```diff
@@ -53,36 +59,52 @@ export function PanelHeaderTitleItems(props: Props) {
+    }),
+    alerting: css({
+      color: theme.colors.error.text,
+      animation: `${alertingPanel} 1s cubic-bezier(1, 0.1, 0.73, 1) 0s infinite alternate`,
```
think we can skip this animation

torkelo commented:
There are a bunch of style issues (and some minor code issues, duplicated styles etc) but happy to fix them after this is merged.

Commit:
allow PanelChrome to be transparent if set from outside; render titleItems directly as a map for less iterations of the whole array

Commit:
Merge branch 'main' into axelav/panel-header-title-items

torkelo commented on packages/grafana-ui/src/components/PanelChrome/PanelChrome.tsx:
```diff
@@ -28,6 +28,7 @@ export interface PanelChromeProps {
   height: number;
   children: (innerWidth: number, innerHeight: number) => ReactNode;
   padding?: PanelPadding;
+  transparent?: boolean;
```
@polibb ah, was hoping this was going to be added in a separate PR. 

I was thinking of a property named background with options `transparent` | 'primary'   (default to primary) is preferrable. I can see a future where we want to support different backgrounds (potentially panels with secondary background, or some other background)

torkelo replied:
@polibb 

or maybe background is not the best name as it also controls border.

displayMode?: 'default' | 'transparent' ; 

naming is hard :)

polibb replied:
@torkelo I see, this is better off in a separate PR then. The only reason I added it is because I thought that's the only case we're covering and it's an easy change. But in a separate PR we can think about the potential use cases.

torkelo commented on public/app/features/dashboard/dashgrid/PanelLinks.tsx:
```diff
@@ -1,39 +1,77 @@
+    menuTrigger: css({
       border: 'none',
       borderRadius: theme.shape.borderRadius(0),
+      cursor: 'context-menu',
+    }),
+    singleLink: css({
+      color: theme.colors.text.secondary,
+      padding: `${theme.spacing(0, 1)}`,
+      height: ` ${theme.spacing(theme.components.height.md)}`,
+      display: 'flex',
+      alignItems: 'center',
+      justifyContent: 'center',
+
+      '&:focus, &:focus-visible': {
+        ...getFocusStyles(theme),
+        zIndex: 1,
+      },
+      '&: focus:not(:focus-visible)': getMouseFocusStyles(theme),
+
+      '&:hover ': {
+        boxShadow: `${theme.shadows.z1}`,
+        color: `${theme.colors.text.primary}`,
+        background: `${theme.colors.background.secondary}`,
+      },
```
we should refactor this out to a shared component as it's duplicated a bit, but we can do that in a follow-up PR

Commit:
remove transparent prop

Commit:
do not animate alerting heart signal

Commit:
no animation on alerting icon

Commit:
fix title items tests

axelavargas committed:
Remove squared outline focus and remove tabIndex in Tooltip component,
now is focusable by default

axelavargas committed:
Merge branch 'main' into axelav/panel-header-title-items

axelavargas committed:
Merge branch 'main' into axelav/panel-header-title-items
